nwb_conversion_tools/utils/specdatachunkiterator.py

```diff
@@ -0,0 +1,74 @@
+import numpy as np
+
+from hdmf.data_utils import AbstractDataChunkIterator, DataChunk
+
+
+class SpecDataChunkIterator(AbstractDataChunkIterator):
+    """DataChunkIterator that lets the user specify chunk shapes."""
+
+    def __init__(self, data: Iterable, full_shape: tuple, chunk_shape: tuple, dtype: str):
+        self.data = data
+        self.full_shape = full_shape
+        self.chunk_shape = chunk_shape
+        self.__curr_timestep = 0
+        self.__curr_channel_idx = 0
+        self.__dtype = DataChunk(data=np.empty(shape=(0, 0), dtype=dtype))
+
+    def __iter__(self):
+        """Return the iterator object"""
+        return self
+
+    def __read_block(self, start_time_idx, stop_time_idx, start_ch_idx, stop_ch_idx):
+        arr = self.data[start_time_idx:stop_time_idx, start_ch_idx:stop_ch_idx]
+        return arr
+
+    def __next__(self):
+        """
+        Return the next data chunk or raise a StopIteration exception if all chunks have been retrieved.
+        :returns: DataChunk object with the data and selection of the current chunk
+        :rtype: DataChunk
+        """
+        end_timestep = self.__curr_timestep + self.chunk_shape[0]
+        if self.__curr_channel_idx >= self.full_shape[1]:
+            self.__curr_channel_idx = 0
+            self.__curr_timestep = end_timestep
+            end_timestep = self.__curr_timestep + self.chunk_shape[0]
+
+            if self.__curr_timestep >= self.full_shape[0]:
+                raise StopIteration
+
+        end_channel_idx = self.__curr_channel_idx + self.chunk_shape[1]
+        if end_timestep > self.full_shape[0]:
+            end_timestep = self.full_shape[0]
+        if end_channel_idx > self.full_shape[1]:
+            end_channel_idx = self.full_shape[1]
+        selection = np.s_[self.__curr_timestep:end_timestep, self.__curr_channel_idx:end_channel_idx]
+        data = self.__read_block(self.__curr_timestep, end_timestep, self.__curr_channel_idx, end_channel_idx)
+        self.__curr_channel_idx = end_channel_idx
+
+        return DataChunk(data=data, selection=selection)
+
+    def recommended_data_shape(self):
+        """
+        Recommend the initial shape for the data array.
+        This is useful in particular to avoid repeated resized of the target array when reading from
+        this data iterator. This should typically be either the final size of the array or the known
+        minimal shape of the array.
+        :return: NumPy-style shape tuple indicating the recommended initial shape for the target array.
+                 This may or may not be the final full shape of the array, i.e., the array is allowed
+                 to grow. This should not be None.
+        """
+        return (self.full_shape[0], self.full_shape[1])
+
+    @property
+    def dtype(self):
+        return np.dtype("int16")
+
+    @property
+    def maxshape(self):
+        return self.full_shape
+
+    def recommended_chunk_shape(self):
+        return None
```

Comment:
Note, in HDF5DATIO to ``recommended_chunk_shape`` is used to determine the chunk settings of the HDF5 dataset https://github.com/hdmf-dev/hdmf/blob/c5f1bf7e476ce6538726db5fa84441c9ddde9957/src/hdmf/backends/hdf5/h5tools.py#L1336-L1338  When returning ``None`` here this is the same as setting ``chunking=True`` in h5py, i.e., h5py will determine which chunk shape to use. That is, unless you are using this in combination with ``H5DataIO`` to explicitly set a chunking, h5py will pick a chunking for you. To ensure reasonable write performance a key piece to keep in mind is that the chunks you write should ideally line up with the chunks of the HDF5 dataset, i.e., usually you want your ``DataChunk`` to cover a set of consecutive chunks in the HDF5 file (or at least cover one full chunk). Part of the issue is that if your ``DataChunk`` and HDF5 chunks don't line up, that on write you will update each HDF5 chunk multiple times. If the HDF5 chunk is no longer in cache, this means HDF5 will have to: 1) read the chunk, 2) apply any filters (e.g., to decompress the chunk), 3) update/merge the data values with the new data, 4) apply any filters (e.g,. compression), 5) write the chunk back to disk. This means, if the chunk you write don't line up with the chunks on disk you potentially end up with a lot more data movement and I/O than you might expect. Long story short, if you allow the user to specify a the chunk_shape used to write the data you may also want to use a recommended shape that aligns well with the given chunk shape.

Reply:
Fantastic info, thanks! I haven't dug into the HDF5DataIO part of it yet, but it sounds like I should in this case to ensure the code is as streamlined as possible.

Reply:
Maybe we can just have the user set `chunk_shape` and `n_chunks`, where `n_chunks` is the number of chunks read and written on each iteration. This would lock the user to reading/writing in a pattern that fits perfectly into the chunking pattern. Another option is to have the user set `buffer_mb` and automatically determine `n_chunks` from that parameter, approximating the desired `buffer_mb` with multiples of chunks, rounding down. The read/write pattern here would still follow the chunking pattern.

Reply:
You can have a look at how h5py does the autochunking here https://github.com/h5py/h5py/blob/master/h5py/_hl/filters.py#L334-L389 to see how h5py guesses chunk sizes. For you the logic should be much simpler since you'd want the chunk-shape for I/O to be a multiple of ``self.chunk_shape``. I.e., the you can probably just iteratively multiply/divide the dimensions of ``self.chunk_shape`` by a factor of 2 until you reach the desired size. Something like this might work:

```
def recommended_chunk_shape(self):
    def get_chunksize(chunks, typesize):
        # Helper function to determine the size in bytes of a chunk
        return np.product(chunks)*typesize    
    # Start with recommending self.chunks as the chunk size
    chunks = [c for c in self.chunks] 
    typesize = self.dtype.itemsize
    target_max_size = h5py.filters.CHUNK_MAX
    target_min_size = h5py.filters.CHUNK_MIN

    # I/O chunks are too large so we need to shrink them
    if get_chunksize(chunks, typesize) > target_max_size: 
         iter_idx = 0
         while (get_chunksize(chunks, typesize) > target_max_size) and  np.product(chunks) > 2:
              chunks[iter_idx%ndims] = np.ceil(chunks[iter_idx%ndims] / 2.0)
              iter_idx += 1
    # I/O chunks are too small so we should grow them
    elif get_chunksize(chunks, typesize) < target_min_size:
         iter_idx = 0
         while (get_chunksize(chunks, typesize) < target_min_size):
              chunks[iter_idx%ndims] = np.ceil(chunks[iter_idx%ndims] * 2.0)
              iter_idx += 1
  
    return tuple(int(x) for x in chunks)
```

Now, a user can always overwrite this behavior by wrapping your iterator in a ``H5DataIO`` object and setting the chunking option explicitly, e.g., if they want to set a chunking that better aligns with typical read patterns rather than the write.

Reply:
The `buffer_mb` you're suggesting would definitely be more intuitive for the user to set, I'll see if I can get the data sizing calculation down.

Reply:
One thing I'm finding with this implementation of the chunk shape inference is overcorrection along the axes if one dimension is at a different magnitude than the other.

E.g., if writing very few long time series, like `full_shape=(2e9, 400)`, the chunk_shape will be pushed to `(x, 1)` where `x` is the number of time frames that can fit into the chunk, whereas we might instead want it contained to `(x/n, n)` where n is some minimal number of series to always chunk together.

On the flip side, if it's a short-ish time series (and the iterations complete with high probability within 3-4 steps), I find that it under-chunks the second dimension by only dividing them by at most 1/4.

I guess in the end it really comes down to compiling all the reasons we want full control of the chunk shape and trying to figure out what is most optimal in the end. As I understand, the compression works on a chunk-to-chunk basis, so different shapes can produce different extents of that. And if we're loading data in for visualizations like in the widgets, that's also sensitive to how the data you're reading relates to the shapes of the stored chunks?

I think I remember hearing that the `ros3` streaming was also done on a chunk to chunk basis? If so that's something that would definitely benefit from a cost-benefit using average download speeds.